monai/transforms/croppad/functional.py

```diff
@@ -0,0 +1,110 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A collection of "vanilla" transforms for spatial operations
+https://github.com/Project-MONAI/MONAI/wiki/MONAI_Design
+"""
+
+from __future__ import annotations
+
+import numpy as np
+import torch
+from torch.nn.functional import pad as pad_pt
+
+from monai.data.meta_obj import get_track_meta
+from monai.data.meta_tensor import MetaTensor
+from monai.transforms.inverse import TraceableTransform
+from monai.transforms.utils import convert_pad_mode, create_translate
+from monai.utils import TraceKeys, convert_to_dst_type, convert_to_tensor
+
+__all__ = ["pad_func"]
+
+
+def pad_func(img: torch.Tensor, to_pad: list[tuple[int, int]], mode: str, transform_info: dict, kwargs: dict):
+    """
+    Args:
+        img: data to be transformed, assuming `img` is channel-first and padding doesn't apply to the channel dim.
+        to_pad: the amount to be padded in each dimension [(low_H, high_H), (low_W, high_W), ...].
+            default to `self.to_pad`.
+        mode: available modes: (Numpy) {``"constant"``, ``"edge"``, ``"linear_ramp"``, ``"maximum"``,
+            ``"mean"``, ``"median"``, ``"minimum"``, ``"reflect"``, ``"symmetric"``, ``"wrap"``, ``"empty"``}
+            (PyTorch) {``"constant"``, ``"reflect"``, ``"replicate"``, ``"circular"``}.
+            One of the listed string values or a user supplied function. Defaults to ``"constant"``.
+            See also: https://numpy.org/doc/1.18/reference/generated/numpy.pad.html
+            https://pytorch.org/docs/stable/generated/torch.nn.functional.pad.html
+        transform_info: a dictionary with the relevant information pertaining to an applied transform.
+        kwargs: other arguments for the `np.pad` or `torch.pad` function.
+            note that `np.pad` treats channel dimension as the first dimension.
+    """
+
+    def _np_pad(img: torch.Tensor, pad_width: list[tuple[int, int]], mode: str, **kwargs) -> torch.Tensor:
```

Comment:
This makes the functions not reachable by the users, is it really necessary to hide? Ateast I see pad_nd is a good utility. Previously the user can call Pad.pad_nd

Reply:
Can we set default value for `transform_info` and `kwargs` like this `, transform_info: dict = {}, kwargs: dict = {}`? Then users could directly call `pad_func`.

Reply:
I don't think they are the same thing, the utilities can keep the data types, right?

Reply:
Oh, I see.

Reply:
Or we could add a flag in `pad_func`, if true, just return `pad_nd`? I can't find another way to avoid `A` <-> `B` logic.

Reply:
what is A<->B, do you mean `pad_nd` <-> `pad_func`?

Reply:
We can have a few functions here:


def _np_pad():
  ...

def _pt_pad():
  ...

def pad_nd():
  ...

def pad_func():
  ...

It’s also fine if you want to put these functions elesewhere. But please make pad_nd and pad_func available to the users. pad_nd is a useful low level implementation, pad_func is the primary functional used by the array transform.

Reply:
Have moved `_np_pad`, `_pt_pad`, and `pad_nd` out of `pad_func` as you suggested!